src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -436,8 +548,10 @@ DatabasePtr DatabaseCatalog::detachDatabase(ContextPtr local_context, const Stri
 
     DatabasePtr db;
     {
+        DatabaseNameHints hints(*this);
+        std::vector<String> names = hints.getHints(database_name, hints.getAllRegisteredNames());
         std::lock_guard lock{databases_mutex};
-        assertDatabaseExistsUnlocked(database_name);
+        assertDatabaseExistsUnlocked(database_name, names);
         db = databases.find(database_name)->second;
```

Comment:
@yariks5s also try to solve the same problem here and other such places - there are 3 in total as I see...
and also I see inefficiency here `assertDatabaseExistsUnlocked` does the same thing as next line - `databases.find` - probably it's time to fix it...

Reply:
This solution solves nothing - the concern is that we retrieve `names` before exception - if there is no exception then `names` is not necessary, and retrieving it is a heavy operation. We should do it only if exception happened. But the problem here is that we already will be in lock on `databases_mutex`, and `getHints` will try to lock it again. The solution is to introduce `DatabaseCatalog::getDatabasesUnlocked()` without lock on `databases_mutex` along with `DatabaseNameHints::getAllRegisteredNamesUnlocked()`, and then call to this inside exception processing in the `assertDatabaseExistsUnlocked` function.
The second optimization I would like to have - modify `assertDatabaseExistsUnlocked` as that it returns iterator from `databases.find(database_name)` call, then you can modify places like
```
assertDatabaseExistsUnlocked(database_name, names);
db = databases.find(database_name)->second;
```
into
```
db = assertDatabaseExistsUnlocked(database_name, names)->second;
```
saving one search. There are at least two of them... maybe more...